contracts/simple_airdrop/src/contract.rs

```diff
@@ -0,0 +1,387 @@
+            ));
+        }
+        config.to_timestamp = to_timestamp;
+        attributes.push(attr("new_to_timestamp", to_timestamp.to_string()))
+    }
+
+    if config.to_timestamp <= config.from_timestamp {
+        return Err(StdError::generic_err("Invalid airdrop claim window"));
+    }
+
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::new().add_attributes(attributes))
+}
+
+/// @dev Facilitates increasing ASTRO airdrop amount
+pub fn handle_increase_astro_incentives(
+    deps: DepsMut,
+    amount: Uint128,
+) -> Result<Response, StdError> {
+    let mut state = STATE.load(deps.storage)?;
+    state.total_airdrop_size += amount;
+    state.unclaimed_tokens += amount;
+
+    STATE.save(deps.storage, &state)?;
+    Ok(Response::new()
+        .add_attribute("action", "astro_airdrop_increased")
+        .add_attribute("total_airdrop_size", state.total_airdrop_size))
+}
+
+/// @dev Executes an airdrop claim for a Terra User
+/// @param claim_amount : Airdrop to be claimed by the user
+/// @param merkle_proof : Array of hashes to prove the input is a leaf of the Merkle Tree
+/// @param root_index : Merkle Tree root identifier to be used for verification
+pub fn handle_claim(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    claim_amount: Uint128,
+    merkle_proof: Vec<String>,
+    root_index: u32,
+) -> Result<Response, StdError> {
+    let recipient = info.sender;
+
+    let config = CONFIG.load(deps.storage)?;
+    let mut state = STATE.load(deps.storage)?;
+
+    // CHECK :: IS AIRDROP CLAIM WINDOW OPEN ?
+    if config.from_timestamp > env.block.time.seconds() {
+        return Err(StdError::generic_err("Claim not allowed"));
+    }
+
+    // CHECK :: IS AIRDROP CLAIM WINDOW OPEN ?
+    if config.to_timestamp < env.block.time.seconds() {
+        return Err(StdError::generic_err("Claim period has concluded"));
+    }
+
+    let merkle_root = config.merkle_roots.get(root_index as usize);
+    if merkle_root.is_none() {
+        return Err(StdError::generic_err("Incorrect Merkle Root Index"));
+    }
+
+    if !verify_claim(&recipient, claim_amount, merkle_proof, merkle_root.unwrap()) {
+        return Err(StdError::generic_err("Incorrect Merkle Proof"));
+    }
+
+    let mut user_info = USERS.load(deps.storage, &recipient).unwrap_or_default();
+
+    // Check if addr has already claimed the tokens
+    if !user_info.airdrop_amount.is_zero() {
+        return Err(StdError::generic_err("Already claimed"));
+    }
+
+    let mut messages = vec![];
+
+    // check is sufficient ASTRO available
+    if state.unclaimed_tokens < claim_amount {
+        return Err(StdError::generic_err("Insufficient ASTRO available"));
+    }
+
+    // TRANSFER ASTRO to the user
+    messages.push(build_transfer_cw20_token_msg(
+        recipient.clone(),
+        config.astro_token_address.to_string(),
+        claim_amount,
+    )?);
+
+    // Update amounts
+    state.unclaimed_tokens -= claim_amount;
```

Comment:
Maybe let's query the balance of tokens for this contract instead of storing amount of unclaimed tokens?

Reply:
i added query balance in `TransferUnclaimedTokens`, should i add it here too and remove the `unclaimed_tokens` balance (right now `unclaimed_tokens` basically tracks how many ASTRO were sent for the airdrop and how many have been claimed from those )

Reply:
yeah, could be a good idea